### ISSUE TITLE
[FEATURE] OnnxScoringModel similar to OnnxEmbeddingModel 

### ISSUE THREAD

**Issue by hzkitty** [langchain4j/langchain4j#1549]
The reordering model ScoringModel currently only supports two types of online models. Can it be similar to OnnxEmbeddingModel to implement the use of OnnxScoringModel and local reordering models

**Comment by hzkitty** [comment]
I wrote a similar one modeled after OnnxEmbeddingModel and tested it normally using the local bge-reranker-large, please feel free to correct me

public class OnnxScoringModel extends AbstractInProcessScoringModel {

    private final OnnxScoringBertBiEncoder onnxBertBiEncoder;

    public OnnxScoringModel(String pathToModel, String pathToTokenizer) {
        this.onnxBertBiEncoder = loadFromFileSystem(pathToModel, pathToTokenizer);
    }

    protected OnnxScoringBertBiEncoder model() {
        return this.onnxBertBiEncoder;
    }
}

import dev.langchain4j.data.segment.TextSegment;
import dev.langchain4j.model.output.Response;
import dev.langchain4j.model.scoring.ScoringModel;

import java.util.ArrayList;
import java.util.List;

public abstract class AbstractInProcessScoringModel implements ScoringModel {

    public AbstractInProcessScoringModel() {
    }

    static OnnxScoringBertBiEncoder loadFromFileSystem(String pathToModel, String pathToTokenizer) {
        try {
            return new OnnxScoringBertBiEncoder(pathToModel, pathToTokenizer);
        } catch (Exception var4) {
            throw new RuntimeException(var4);
        }
    }

    protected abstract OnnxScoringBertBiEncoder model();

    public Response<List<Double>> scoreAll(List<TextSegment> segments, String query) {
        List<Double> scores = new ArrayList<>();

        double score;
        for (TextSegment segment : segments) {
            score = this.model().score(segment.text(), query);
            scores.add(score);
        }
        return Response.from(scores);
    }
}

import ai.djl.huggingface.tokenizers.Encoding;
import ai.djl.huggingface.tokenizers.HuggingFaceTokenizer;
import ai.onnxruntime.OnnxTensor;
import ai.onnxruntime.OrtEnvironment;
import ai.onnxruntime.OrtException;
import ai.onnxruntime.OrtSession;

import java.nio.LongBuffer;
import java.nio.file.Paths;
import java.util.*;

public class OnnxScoringBertBiEncoder {
    private final OrtEnvironment environment;
    private final OrtSession session;
    private final HuggingFaceTokenizer tokenizer;

    public OnnxScoringBertBiEncoder(String modelPath, String pathToTokenizer) {
        try {
            this.environment = OrtEnvironment.getEnvironment();
            this.session = this.environment.createSession(modelPath);
            this.tokenizer = HuggingFaceTokenizer.newInstance(Paths.get(pathToTokenizer), Map.of("truncation", "LONGEST_FIRST")); // LONGEST_FIRST 优先截断最长的那部分
        } catch (Exception var4) {
            throw new RuntimeException(var4);
        }
    }

    public double score(String text, String query) {
        double score;
        try {
            OrtSession.Result result = this.encode(text, query);
            Throwable var8 = null;

            try {
                score = this.toScore(result);
            } catch (Throwable var18) {
                var8 = var18;
                throw var18;
            } finally {
                if (result != null) {
                    if (var8 != null) {
                        try {
                            result.close();
                        } catch (Throwable var17) {
                            var8.addSuppressed(var17);
                        }
                    } else {
                        result.close();
                    }
                }

            }
        } catch (OrtException var20) {
            throw new RuntimeException(var20);
        }

        return score;
    }

    private OrtSession.Result encode(String text, String query) throws OrtException {
        String[] pairs = new String[]{query, text};
        Encoding encoding = this.tokenizer.encode(pairs, true, false);
        long[] inputIds = encoding.getIds();
        long[] attentionMask = encoding.getAttentionMask();
        long[] shape = new long[]{1L, (long)inputIds.length};
        OnnxTensor inputIdsTensor = OnnxTensor.createTensor(this.environment, LongBuffer.wrap(inputIds), shape);
        Throwable var8 = null;

        OrtSession.Result var14;
        try {
            OnnxTensor attentionMaskTensor = OnnxTensor.createTensor(this.environment, LongBuffer.wrap(attentionMask), shape);
            Throwable var10 = null;
            try {
                Map<String, OnnxTensor> inputs = new HashMap();
                inputs.put("input_ids", inputIdsTensor);
                inputs.put("attention_mask", attentionMaskTensor);

                var14 = this.session.run(inputs);
            } catch (Throwable var60) {
                var10 = var60;
                throw var60;
            } finally {
                if (attentionMaskTensor != null) {
                    if (var10 != null) {
                        try {
                            attentionMaskTensor.close();
                        } catch (Throwable var56) {
                            var10.addSuppressed(var56);
                        }
                    } else {
                        attentionMaskTensor.close();
                    }
                }

            }
        } catch (Throwable var62) {
            var8 = var62;
            throw var62;
        } finally {
            if (inputIdsTensor != null) {
                if (var8 != null) {
                    try {
                        inputIdsTensor.close();
                    } catch (Throwable var55) {
                        var8.addSuppressed(var55);
                    }
                } else {
                    inputIdsTensor.close();
                }
            }
        }
        return var14;
    }

    private double toScore(OrtSession.Result result) throws OrtException {
        float[][] output = (float[][]) result.get(0).getValue();
        return this.sigmoid(output[0][0]);
    }

    private double sigmoid(float x) {
        return 1 / (1 + Math.exp(-x));
    }
}

**Comment by langchain4j** [comment]
Related: https://github.com/langchain4j/langchain4j-embeddings/compare/main...cross_encoders

**Comment by aseovic** [comment]
I tried the code above, and unfortunately it doesn't seem to work. 

When I try to run it I get the following error:
```
ai.onnxruntime.OrtException: Error code - ORT_RUNTIME_EXCEPTION - message: 
Non-zero status code returned while running Gather node. Name:'/embeddings/token_type_embeddings/Gather' 
Status Message: /Users/runner/work/1/s/include/onnxruntime/core/framework/op_kernel_context.h:42 const T *onnxruntime::OpKernelContext::Input(int) const [T = onnxruntime::Tensor] 
Missing Input: token_type_ids
```

Unfortunately, I don't know enough about ONNX, tensors, etc. to understand what's going on and how to create the missing input and help fix this, so hopefully someone who knows more about all of this can take a look.

**Comment by langchain4j** [comment]
@aseovic which code? The one from @1758225523 or https://github.com/langchain4j/langchain4j-embeddings/compare/main...cross_encoders ?

**Comment by aseovic** [comment]
The one from @1758225523. 

I was actually able to get `OnnxBertCrossEncoder` from the `cross_encoders` link to get to work, with some minor tweaks (basically, the addition of `score` and `sigmoid` methods):

```
    public double score(String query, String text)
        {
        return sigmoid(encode(query, text));
        }

    private double sigmoid(float x)
        {
        return 1 / (1 + Math.exp(-x));
        }
```

**Comment by langchain4j** [comment]
@aseovic great! Is the output the same as python libs produce?
Would you lile to finish that work and submit a PR?

**Comment by hzkitty** [comment]
Updated version

import ai.onnxruntime.OrtSession;

public class OnnxScoringModel extends AbstractInProcessScoringModel {

    private final OnnxScoringBertBiEncoder onnxBertBiEncoder;

    public OnnxScoringModel(String pathToModel, OrtSession.SessionOptions option, String pathToTokenizer) {
        this.onnxBertBiEncoder = loadFromFileSystem(pathToModel, option, pathToTokenizer);
    }

    protected OnnxScoringBertBiEncoder model() {
        return this.onnxBertBiEncoder;
    }
}


import ai.onnxruntime.OrtSession;
import dev.langchain4j.data.segment.TextSegment;
import dev.langchain4j.model.output.Response;
import dev.langchain4j.model.scoring.ScoringModel;

import java.util.List;

public abstract class AbstractInProcessScoringModel implements ScoringModel {
    
    public AbstractInProcessScoringModel() {
    }

    static OnnxScoringBertBiEncoder loadFromFileSystem(String pathToModel, OrtSession.SessionOptions option, String pathToTokenizer) {
        try {
            return new OnnxScoringBertBiEncoder(pathToModel, option, pathToTokenizer);
        } catch (Exception var4) {
            throw new RuntimeException(var4);
        }
    }

    protected abstract OnnxScoringBertBiEncoder model();

    public Response<List<Double>> scoreAll(List<TextSegment> segments, String query) {
        List<Double> scores = this.model().scoreAll(query, segments.stream().map(TextSegment::text).toList());
        return Response.from(scores);
    }
}

import ai.djl.huggingface.tokenizers.Encoding;
import ai.djl.huggingface.tokenizers.HuggingFaceTokenizer;
import ai.djl.util.PairList;
import ai.onnxruntime.OnnxTensor;
import ai.onnxruntime.OrtEnvironment;
import ai.onnxruntime.OrtException;
import ai.onnxruntime.OrtSession;
import ai.onnxruntime.OrtSession.Result;

import java.nio.file.Paths;
import java.util.*;

import static ai.onnxruntime.OnnxTensor.createTensor;

public class OnnxScoringBertBiEncoder {

    private static final int MAX_SEQUENCE_LENGTH = 510; // 512 - 2 (special tokens [CLS] and [SEP]) 根据你模型支持的最大token调整

    private final OrtEnvironment environment;
    private final OrtSession session;
    private final Set<String> expectedInputs;
    private final HuggingFaceTokenizer tokenizer;

    public OnnxScoringBertBiEncoder(String modelPath, OrtSession.SessionOptions options, String pathToTokenizer) {
        try {
            this.environment = OrtEnvironment.getEnvironment();
            this.session = this.environment.createSession(modelPath, options);
            this.expectedInputs = session.getInputNames();
            Map<String, String> tokenizerOptions = Map.of("padding", "true","truncation", "LONGEST_FIRST", "modelMaxLength", String.valueOf(MAX_SEQUENCE_LENGTH));
            this.tokenizer = HuggingFaceTokenizer.newInstance(Paths.get(pathToTokenizer), tokenizerOptions); // 默认最大长度限制，LONGEST_FIRST 优先截断最长的那部分
        } catch (Exception var4) {
            throw new RuntimeException(var4);
        }
    }

    public List<Double> scoreAll(String query, List<String> documents) {
        List<Double> scores;
        // 创建输入列表
        PairList<String, String> pairs = new PairList<>();
        for (String document : documents) {
            pairs.add(query, document);
        }
        try (Result result = this.encode(pairs)) {
            scores = this.toScore(result);
        } catch (OrtException e) {
            throw new RuntimeException(e);
        }
        return scores;
    }

    private Result encode(PairList<String, String> pairs) throws OrtException {
        // 对每一对句子进行编码
        Encoding[] encodings = this.tokenizer.batchEncode(pairs);
        long[][] inputIds = new long[encodings.length][];
        long[][] attentionMask = new long[encodings.length][];
        long[][] tokenTypeIds = new long[encodings.length][];
        for (int i = 0; i < encodings.length; i++) {
            inputIds[i] = encodings[i].getIds();
            attentionMask[i] = encodings[i].getAttentionMask();
            tokenTypeIds[i] = encodings[i].getTypeIds();
        }

        try (
                OnnxTensor inputIdsTensor = createTensor(environment, inputIds);
                OnnxTensor attentionMaskTensor = createTensor(environment, attentionMask);
                OnnxTensor tokenTypeIdsTensor = createTensor(this.environment, tokenTypeIds);
        ) {
            Map<String, OnnxTensor> inputs = new HashMap<>();
            inputs.put("input_ids", inputIdsTensor);
            inputs.put("attention_mask", attentionMaskTensor);

            if (expectedInputs.contains("token_type_ids")) {
                inputs.put("token_type_ids", tokenTypeIdsTensor);
            }

            return session.run(inputs);
        }
    }

    private List<Double> toScore(Result result) throws OrtException {
        float[][] output = (float[][]) result.get(0).getValue();
        List<Double> scores = new ArrayList<>();
        for (int i = 0; i < output.length; i++) {
            scores.add(sigmoid(output[i][0]));
        }
        return scores;
    }

    private double sigmoid(float x) {
        return 1 / (1 + Math.exp(-x));
    }

}

**Comment by aseovic** [comment]
@langchain4j I don't know if it matches Python results, but I can check tomorrow.

On the surface it looks pretty good:

Model: **`ms-marco-MiniLM-L-6-v2`**
Question: **What is panda?**
Scored Answers:
```
0.00003: Dolphins are mammals, not fish.
0.99973: The giant panda (Ailuropoda melanoleuca), sometimes called a panda bear or simply panda, is a bear species endemic to China.
```

Model: **`ms-marco-TinyBERT-L-2-v2`**
Question: **How many people live in Berlin?**
Scored Answers:
```
0.00001: New York City is famous for the Metropolitan Museum of Art.
0.99922: Berlin had a population of 3,520,031 registered inhabitants in an area of 891.82 square kilometers.
```

I'll do the same test in Python tomorrow and will post the results.

**Comment by aseovic** [comment]
@1758225523  Does the latest code improve encoding performance by batching multiple pairs into a single model request?

I'm happy to give that a try as well.

The code I have now is somewhat of a merge of first two approaches. It's mostly the original `OnnxBertCrossEncoder`, but modified to use `HuggingFaceTokenizer` instead of deprecated `BertTokenizer`:

```java
    private Result run(String query, String text) throws OrtException
        {
        Encoding encoding = tokenizer.encode(query, text, true, false);

        long[] inputIds      = encoding.getIds();
        long[] attentionMask = encoding.getAttentionMask();
        long[] tokenTypeIds  = encoding.getTypeIds();
        long[] shape         = new long[] {1L, inputIds.length};

        try (OnnxTensor tokensTensor = createTensor(environment, wrap(inputIds), shape);
             OnnxTensor attentionMasksTensor = createTensor(environment, wrap(attentionMask), shape);
             OnnxTensor tokenTypeIdsTensor = createTensor(environment, wrap(tokenTypeIds), shape))
            {
            Map<String, OnnxTensor> inputs = new HashMap<>();
            inputs.put("input_ids", tokensTensor);
            inputs.put("token_type_ids", tokenTypeIdsTensor);
            inputs.put("attention_mask", attentionMasksTensor);

            return session.run(inputs);
            }
        }
```

**Comment by aseovic** [comment]
@langchain4j To answer your other question, sure, happy to create a PR with whatever a final solution is, unless @1758225523 wants to do it (it's more his/her work than mine...)

**Comment by langchain4j** [comment]
@aseovic scores look very extreme TBH, is it due to `sigmoid` change?

**Comment by aseovic** [comment]
Yes, `sigmoid` effectively "normalizes" the result to a value between 0 and 1.

Without it, the results are:

Model: **`ms-marco-MiniLM-L-6-v2`**
Question: **What is panda?**
Scored Answers:
```
-10.522052 (-10.522055): Dolphins are mammals, not fish.
  8.221886 (  8.221886): The giant panda (Ailuropoda melanoleuca), sometimes called a panda bear or simply panda, is a bear species endemic to China.
```

Model: **`ms-marco-TinyBERT-L-2-v2`**
Question: **How many people live in Berlin?**
Scored Answers:
```
-11.562293 (-11.562296): New York City is famous for the Metropolitan Museum of Art.
  7.152364 (  7.152371): Berlin had a population of 3,520,031 registered inhabitants in an area of 891.82 square kilometers.
```
The results in parenthesis are Python results, so I'd say we are good ;-)

**Comment by langchain4j** [comment]
@aseovic great! I would return original scores, user can always apply sigmoid if required

**Comment by aseovic** [comment]
Yup, that's what I was thinking as well 👍
